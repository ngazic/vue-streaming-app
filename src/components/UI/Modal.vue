<template>
<!-- Modal -->
<div class="modal-dialog modal-dialog-centered">
  <div class="modal-content">
    <div class="modal-header">
      <h5 class="modal-title" id="staticBackdropLabel">Modal title</h5>
      <button type="button" class="close" data-dismiss="modal" aria-label="Close" @click="cancel">
        <span aria-hidden="true">&times;</span>
      </button>
    </div>
    <div class="modal-body">
      This is demo payment option. Clicking "Understood" will simulate payment service, and you will proceed with watching desired video
    </div>
    <div class="modal-footer">
      <button type="button" class="btn btn-secondary" data-dismiss="modal" @click="cancel">
        Close
      </button>
      <button type="button" class="btn btn-primary" @click="agreed">Understood</button>
    </div>
  </div>
</div>
</template>

<script lang="ts">
export default {
  methods: {
    agreed(): void {
      this.$emit('handlePayment', true)
    },
    cancel(): void {
      this.$emit('handlePayment', false)
    }
  }
};
</script>
