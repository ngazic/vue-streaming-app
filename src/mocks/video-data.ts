export interface Video {
  id: number;
  title: string;
  url: string;
  poster: string;
  price: number;
}

function priceGenerator(): number {
  return Math.floor(Math.random() * 30);
}

export const allVideos = [
  {
id: 1,
title: 'Video 1',
url: '',
poster: 'https://loremflickr.com/320/240/movies/all?1',
price: priceGenerator()
  },
  {
id: 2,
title: 'Video 2',
url: 'https://ak.picdn.net/shutterstock/videos/1054068659/preview/stock-footage-group-of-young-boys-play-soccer-training-day-on-the-football-field-teenagers-play-football-on-a.webm',
poster: 'https://loremflickr.com/320/240/movies/all?3',
price: priceGenerator()
  },
  {
id: 3,
title: 'Video 3',
url: 'https://ak.picdn.net/shutterstock/videos/1054068659/preview/stock-footage-group-of-young-boys-play-soccer-training-day-on-the-football-field-teenagers-play-football-on-a.webm',
poster: 'https://loremflickr.com/320/240/movies/all?4',
price: priceGenerator()
  },
  {
id: 4,
title: 'Video 4',
url: 'https://ak.picdn.net/shutterstock/videos/1054068659/preview/stock-footage-group-of-young-boys-play-soccer-training-day-on-the-football-field-teenagers-play-football-on-a.webm',
poster: 'https://loremflickr.com/320/240/movies/all?8',
price: priceGenerator()
  },
  {
id: 5,
title: 'Video 5',
url: 'https://ak.picdn.net/shutterstock/videos/1054068659/preview/stock-footage-group-of-young-boys-play-soccer-training-day-on-the-football-field-teenagers-play-football-on-a.webm',
poster: 'https://loremflickr.com/320/240/movies/all?14',
price: priceGenerator()
  },
  {
id: 6,
title: 'Video 6',
url: 'https://ak.picdn.net/shutterstock/videos/1054068659/preview/stock-footage-group-of-young-boys-play-soccer-training-day-on-the-football-field-teenagers-play-football-on-a.webm',
poster: 'https://loremflickr.com/320/240/sport/all',
price: priceGenerator()
  },
  {
id: 7,
title: 'Video 7',
url: 'https://ak.picdn.net/shutterstock/videos/1054068659/preview/stock-footage-group-of-young-boys-play-soccer-training-day-on-the-football-field-teenagers-play-football-on-a.webm',
poster: 'https://loremflickr.com/320/240/fun/all',
price: priceGenerator()
  },
]

export async function getVideos(): Promise<Video[]> {
  return allVideos;
}