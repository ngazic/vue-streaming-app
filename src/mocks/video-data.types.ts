export interface Video {
  id: number;
  title: string;
  url: string;
  poster: string;
  price: number;
}