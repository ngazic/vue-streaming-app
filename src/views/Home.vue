<template>
<div class="home">
  <div class="container-fluid">
    <div v-if="!showModal" class="row">
      <div class="col-12 col-sm-6 col-md-4 col-lg-2" v-for="video in videos" :key="video.id">
        <div class="card" @click="click(video)">
          <div class="card-body">
            <img :src="video.poster" class="card-img-top" />
            <h5 class="card-title">{{ video.title }}</h5>
            <p>Price {{ video.price }}$</p>
          </div>
        </div>
      </div>
    </div>
    <modal v-if="showModal" @handlePayment="payment" />
  </div>
</div>
</template>

<script lang="ts">
import Vue from "vue";
import Modal from "@/components/UI/Modal.vue";
import {
  Video,
} from "../mocks/video-data.types";
import {
  getVideos
} from "../mocks/video-data";
import {
  mapActions
} from 'vuex'
const initialVideoList: Video[] = [];
let initialVideo: Video;
export default Vue.extend({
  name: "Home",
  components: {
    Modal
  },
  async created() {
    this.videos = await getVideos();
  },
  data() {
    return {
      videos: initialVideoList,
      showModal: false,
      paymentVerified: false,
      currentVideo: initialVideo
    };
  },
  methods: {
    ...mapActions(['paymentAction']),
    payment(data: boolean): void {
      // console.log(data);
      if (data) {
        this.paymentVerified = true;
        this.paymentAction(this.currentVideo)
        this.$store;
        this.$router.push("/watch");
      }
      this.showModal = false;
    },
    click(video: Video): void {
      this.showModal = true;
      this.currentVideo = video;
      // console.log(video);
    }
  }
});
</script>
