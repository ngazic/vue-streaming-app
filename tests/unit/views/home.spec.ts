import { shallowMount, createLocalVue, Wrapper } from '@vue/test-utils'
import VueRouter from 'vue-router'
import Vuex from 'vuex'
import Modal from '@/components/UI/Modal.vue'
import Home from '@/views/Home.vue'
import { Video, allVideos } from '@/mocks/video-data'

describe('Home.vue', () => {
  let wrapper: Wrapper<Home>
  let video: Wrapper<Home>
  const initialVideo: Video = {
    id: -1,
    poster: '',
    price: 0,
    title: '',
    url: ''
  }
  const localVue = createLocalVue()
  const actions = {
    paymentAction: () => jest.fn()
  }
  localVue.use(VueRouter)
  localVue.use(Vuex)
  const router = new VueRouter()
  const store = new Vuex.Store({
    actions
  })

  beforeEach(() => {
    wrapper = shallowMount(Home, {
      localVue,
      store,
      router,
      data: () => ({
        videos: allVideos,
        showModal: false,
        paymentVerified: false,
        currentVideo: initialVideo
      })
    })
    video = wrapper.find('.card')
  })

  it('should fetch some data', () => {
    expect(video.exists()).toBe(true)
  });

  it('should show modal on video clicked', async () => {
    await video.trigger('click')
    expect(wrapper.findComponent(Modal).exists()).toBe(true)
  });
  it('should redirect to clicked video if payment is confirmed', () => {
  // console.log(wrapper.vm.payment)
  wrapper.setData({paymentVerified: true});
  console.log(wrapper.vm.$children)
    expect(jest.spyOn(actions,'paymentAction')).toHaveBeenCalledWith({})
  });

  it('dispatch "payment" action with payed video', () => {
    expect(true).toBe(false)

  });

  it('shows modal on video click', () => {

  });

  it('router called to watch payed video', () => {

  });

})