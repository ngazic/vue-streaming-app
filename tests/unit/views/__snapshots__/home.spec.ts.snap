// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue renders correctly 1`] = `
<div class="home">
  <div class="container-fluid">
    <div class="row">
      <div class="col-12 col-sm-6 col-md-4 col-lg-2">
        <div class="card">
          <div class="card-body"><img src="https://loremflickr.com/320/240/movies/all?1" class="card-img-top">
            <h5 class="card-title">Video 1</h5>
            <p>Price 0$</p>
          </div>
        </div>
      </div>
      <div class="col-12 col-sm-6 col-md-4 col-lg-2">
        <div class="card">
          <div class="card-body"><img src="https://loremflickr.com/320/240/movies/all?3" class="card-img-top">
            <h5 class="card-title">Video 2</h5>
            <p>Price 16$</p>
          </div>
        </div>
      </div>
      <div class="col-12 col-sm-6 col-md-4 col-lg-2">
        <div class="card">
          <div class="card-body"><img src="https://loremflickr.com/320/240/movies/all?4" class="card-img-top">
            <h5 class="card-title">Video 3</h5>
            <p>Price 5$</p>
          </div>
        </div>
      </div>
      <div class="col-12 col-sm-6 col-md-4 col-lg-2">
        <div class="card">
          <div class="card-body"><img src="https://loremflickr.com/320/240/movies/all?8" class="card-img-top">
            <h5 class="card-title">Video 4</h5>
            <p>Price 21$</p>
          </div>
        </div>
      </div>
      <div class="col-12 col-sm-6 col-md-4 col-lg-2">
        <div class="card">
          <div class="card-body"><img src="https://loremflickr.com/320/240/movies/all?14" class="card-img-top">
            <h5 class="card-title">Video 5</h5>
            <p>Price 29$</p>
          </div>
        </div>
      </div>
      <div class="col-12 col-sm-6 col-md-4 col-lg-2">
        <div class="card">
          <div class="card-body"><img src="https://loremflickr.com/320/240/sport/all" class="card-img-top">
            <h5 class="card-title">Video 6</h5>
            <p>Price 2$</p>
          </div>
        </div>
      </div>
      <div class="col-12 col-sm-6 col-md-4 col-lg-2">
        <div class="card">
          <div class="card-body"><img src="https://loremflickr.com/320/240/fun/all" class="card-img-top">
            <h5 class="card-title">Video 7</h5>
            <p>Price 2$</p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
